File: goldenflow/cli/init_wizard.py
"""Interactive setup wizard for GoldenFlow config."""
from __future__ import annotations

from pathlib import Path

from rich.console import Console
from rich.prompt import Prompt, Confirm
from rich.table import Table

console = Console()


def run_wizard(data_path: Path | None = None, output_path: Path = Path("goldenflow.yaml")) -> None:
    """Interactive config generation wizard."""
    from goldenflow.connectors.file import read_file
    from goldenflow.engine.profiler_bridge import profile_dataframe
    from goldenflow.engine.selector import select_transforms
    from goldenflow.config.schema import GoldenFlowConfig, TransformSpec
    from goldenflow.config.loader import save_config

    console.print("[bold]GoldenFlow Setup Wizard[/bold]\n")

    # Step 1: Get data file
    if data_path is None:
        path_str = Prompt.ask("Path to your data file")
        data_path = Path(path_str)

    if not data_path.exists():
        console.print(f"[red]File not found: {data_path}[/red]")
        return

    # Step 2: Profile the data
    console.print(f"\n[dim]Profiling {data_path.name}...[/dim]")
    df = read_file(data_path)
    profile = profile_dataframe(df, file_path=str(data_path))

    # Step 3: Show profile summary
    table = Table(title="Column Profile")
    table.add_column("Column", style="cyan")
    table.add_column("Type", style="green")
    table.add_column("Nulls", style="yellow")
    table.add_column("Suggested Transforms", style="magenta")

    column_transforms: dict[str, list[str]] = {}
    for col in profile.columns:
        selected = select_transforms(col)
        transform_names = [t.name for t in selected]
        column_transforms[col.name] = transform_names
        table.add_row(
            col.name,
            col.inferred_type,
            f"{col.null_count} ({col.null_pct:.0%})",
            ", ".join(transform_names[:3]) or "[dim]none[/dim]",
        )

    console.print(table)

    # Step 4: Let user confirm/modify per column
    transforms: list[TransformSpec] = []
    console.print("\n[bold]Configure transforms per column:[/bold]\n")

    for col_name, suggested in column_transforms.items():
        if not suggested:
            continue
        if Confirm.ask(f"  Apply [{', '.join(suggested)}] to [cyan]{col_name}[/cyan]?", default=True):
            transforms.append(TransformSpec(column=col_name, ops=suggested))

    # Step 5: Ask about renames, drops
    renames: dict[str, str] = {}
    if Confirm.ask("\nRename any columns?", default=False):
        while True:
            old = Prompt.ask("  Old name (or 'done')")
            if old.lower() == "done":
                break
            new = Prompt.ask(f"  New name for '{old}'")
            renames[old] = new

    drop: list[str] = []
    if Confirm.ask("Drop any columns?", default=False):
        drop_str = Prompt.ask("  Column names to drop (comma-separated)")
        drop = [c.strip() for c in drop_str.split(",") if c.strip()]

    # Step 6: Save config
    config = GoldenFlowConfig(
        source=str(data_path),
        transforms=transforms,
        renames=renames,
        drop=drop,
    )
    save_config(config, output_path)
    console.print(f"\n[green]Config saved to {output_path}[/green]")
    console.print(f"[dim]Run: goldenflow transform {data_path} -c {output_path}[/dim]")

File: packages/goldenflow-js/src/node/init-wizard.ts
import { createInterface } from "node:readline";
import { readFile } from "./connectors/file.js";
import { profileDataframe } from "../core/engine/profiler-bridge.js";
import { selectTransforms } from "../core/engine/selector.js";
import { saveConfigToString } from "../core/config/loader.js";
import { makeConfig } from "../core/types.js";
import type { TransformSpec } from "../core/types.js";
import { writeFileSync } from "node:fs";

function ask(rl: ReturnType<typeof createInterface>, question: string): Promise<string> {
  return new Promise((resolve) => rl.question(question, resolve));
}

export async function runWizard(dataPath?: string, outputPath = "goldenflow.yaml"): Promise<void> {
  const rl = createInterface({ input: process.stdin, output: process.stdout });

  try {
    console.log("GoldenFlow Setup Wizard\n");

    const filePath = dataPath ?? await ask(rl, "Path to your data file: ");
    console.log(`\nProfiling ${filePath}...`);

    const rows = readFile(filePath);
    const profile = profileDataframe(rows, filePath);

    console.log(`\n${profile.rowCount} rows, ${profile.columnCount} columns\n`);

    const columnTransforms: Record<string, string[]> = {};
    for (const col of profile.columns) {
      const selected = selectTransforms(col);
      const names = selected.map(t => t.name);
      columnTransforms[col.name] = names;
      console.log(`  ${col.name}: ${col.inferredType} | suggested: ${names.slice(0, 3).join(", ") || "none"}`);
    }

    console.log("\nConfigure transforms per column:\n");
    const transforms: TransformSpec[] = [];

    for (const [colName, suggested] of Object.entries(columnTransforms)) {
      if (suggested.length === 0) continue;
      const answer = await ask(rl, `  Apply [${suggested.join(", ")}] to ${colName}? (Y/n) `);
      if (answer.toLowerCase() !== "n") {
        transforms.push({ column: colName, ops: suggested });
      }
    }

    const config = makeConfig({ source: filePath, transforms });
    const yaml = saveConfigToString(config);
    writeFileSync(outputPath, yaml);
    console.log(`\nConfig saved to ${outputPath}`);
    console.log(`Run: goldenflow-js transform ${filePath} -c ${outputPath}`);
  } finally {
    rl.close();
  }
}
